README.md
## WiFiMulti_Generic Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/WiFiMulti_Generic.svg?)](https://www.ardu-badge.com/WiFiMulti_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/WiFiMulti_Generic.svg)](https://github.com/khoih-prog/WiFiMulti_Generic/releases)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/WiFiMulti_Generic.svg)](http://github.com/khoih-prog/WiFiMulti_Generic/issues)

<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>


---
---

## Table of Contents

* [Why do we need this WiFiMulti_Generic library](#why-do-we-need-this-WiFiMulti_Generic-library)
  * [Features](#features)
  * [Currently supported Boards](#currently-supported-boards)
  * [Currently supported WiFi shields/modules](#currently-supported-wifi-shieldsmodules)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [Packages' Patches](#packages-patches)
  * [1. For Adafruit nRF52840 and nRF52832 boards](#1-for-adafruit-nRF52840-and-nRF52832-boards)
  * [2. For Teensy boards](#2-for-teensy-boards)
  * [3. For Arduino SAM DUE boards](#3-for-arduino-sam-due-boards)
  * [4. For Arduino SAMD boards](#4-for-arduino-samd-boards)
      * [For core version v1.8.10+](#for-core-version-v1810)
      * [For core version v1.8.9-](#for-core-version-v189-)
  * [5. For Adafruit SAMD boards](#5-for-adafruit-samd-boards)
  * [6. For Seeeduino SAMD boards](#6-for-seeeduino-samd-boards)
  * [7. For STM32 boards](#7-for-stm32-boards) 
    * [7.1. For STM32 boards to use LAN8720](#71-for-stm32-boards-to-use-lan8720)
    * [7.2. For STM32 boards to use Serial1](#72-for-stm32-boards-to-use-serial1)
  * [8. For RP2040-based boards using Earle Philhower arduino-pico core](#8-for-rp2040-based-boards-using-earle-philhower-arduino-pico-core)
    * [8.1. To use BOARD_NAME](#81-to-use-board_name)
    * [8.2. To avoid compile error relating to microsecondsToClockCycles](#82-to-avoid-compile-error-relating-to-microsecondstoclockcycles)
  * [9. For Portenta_H7 boards using Arduino IDE in Linux](#9-for-portenta_h7-boards-using-arduino-ide-in-linux)
  * [10. For RTL8720DN boards using AmebaD core](#10-for-rtl8720dn-boards-using-amebad-core)
  * [11. For SAMD21 and SAMD51 boards using ArduinoCore-fab-sam core](#11-For-SAMD21-and-SAMD51-boards-using-ArduinoCore-fab-sam-core)
  * [12. For Seeeduino RP2040 boards](#12-For-Seeeduino-RP2040-boards)
  * [13. For Seeeduino nRF52840 boards](#13-For-Seeeduino-nRF52840-boards)
* [Libraries' Patches](#libraries-patches)
   * [1. ArduinoSTL for AVR boards](#1-ArduinoSTL-for-AVR-boards)
* [Examples](#examples)
  * [ 1. WiFiMulti](examples/WiFiMulti)
  * [ 2. WiFiMulti_RP2040W](examples/WiFiMulti_RP2040W) **New**
* [Example WiFiMulti](#example-WiFiMulti)
  * [1. File WiFiMulti.ino](#1-File-WiFiMultiino)
  * [2. File defines.h](#2-File-definesh)
* [Debug Terminal Output Samples](#debug-terminal-output-samples)
  * [ 1. WiFiMulti on RASPBERRY_PI_PICO](#1-WiFiMulti-on-RASPBERRY_PI_PICO)
  * [ 2. WiFiMulti on MBED RaspberryPi Pico](#2-WiFiMulti-on-MBED-RaspberryPi-Pico)
  * [ 3. WiFiMulti on Nano RP2040 Connect](#3-WiFiMulti-on-Nano-RP2040-Connect)
  * [ 4. WiFiMulti on NANO_RP2040_CONNECT](#4-WiFiMulti-on-NANO_RP2040_CONNECT)
  * [ 5. WiFiMulti on ESP8266_NODEMCU_ESP12E](#5-WiFiMulti-on-ESP8266_NODEMCU_ESP12E)
  * [ 6. WiFiMulti on ESP32C3_DEV](#6-WiFiMulti-on-ESP32C3_DEV)
  * [ 7. WiFiMulti on ESP32S3_DEV](#7-WiFiMulti-on-ESP32S3_DEV)
  * [ 8. WiFiMulti on Arduino AVR Mega2560](#8-WiFiMulti-on-Arduino-AVR-Mega2560)
  * [ 9. WiFiMulti on RTL8720DN](#9-WiFiMulti-on-RTL8720DN)
  * [10. WiFiMulti on PORTENTA_H7_M7](#10-WiFiMulti-on-PORTENTA_H7_M7)
  * [11. WiFiMulti on RP2040W](#11-WiFiMulti-on-RP2040W)
* [Debug](#debug)
* [Troubleshooting](#troubleshooting)
* [Issues](#issues)
* [TO DO](#to-do)
* [DONE](#done)
* [Contributions and Thanks](#contributions-and-thanks)
* [Contributing](#contributing)
* [License](#license)
* [Copyright](#copyright)


---
---

### Why do we need this [WiFiMulti_Generic library](https://github.com/khoih-prog/WiFiMulti_Generic)


#### Features

This [**WiFiMulti_Generic library**](https://github.com/khoih-prog/WiFiMulti_Generic) is a simple MultiWiFi library to adapt ESP32/ESP8266 `MultiWiFi` feature of WiFi libraries to all other WiFi modules and libraries. The functions are similar and compatible to those of [`ESP32 MultiWiFi`](https://github.com/espressif/arduino-esp32/tree/master/libraries/WiFi) and [`ESP8266 MultiWiFi`](https://github.com/esp8266/Arduino/tree/master/libraries/ESP8266WiFi) libraries to make life much easier to port sketches from ESP8266/ESP32.


#### Currently Supported Boards

This [**WiFiMulti_Generic library**](https://github.com/khoih-prog/WiFiMulti_Generic) currently supports these following boards:

 1. SAM DUE

 2. SAMD21

  - Arduino: ZERO, MKR, NANO_33_IOT, etc.
  - Adafruit SAMD21 (M0) : ItsyBitsy M0, Feather M0, Feather M0 Express, Metro M0 Express, Circuit Playground Express, Trinket M0, PIRkey, HalloWing M0, Crickit M0, etc.
  - Seeeduino:  LoRaWAN, Zero, Femto M0, XIAO M0, Wio GPS Board, etc.
  
 3. SAMD51

  - Adafruit SAMD51 (M4) : Metro M4, Grand Central M4, ItsyBitsy M4, Feather M44 Express, Trellis M4, Metro M4 AirLift lite, MONSTER M4SK Express, Hallowing EM4 xpress, etc.
  - Seeeduino: Wio Terminal, Grove UI Wireless
  
 4. Teensy (4.1, 4.0, 3.6, 3.5, 3,2, 3.1, 3.0, LC)
 
 5. All STM32F/L/H/G/WB/MP1 with more than 32KB flash memory using stm32duino `Arduino_Core_STM32`
 
 6. AVR Mega1280, 2560, ADK, 32U4, 16U4, etc. using Arduino, Adafruit or Sparkfun core. To use patch for `ArduinoSTL` library. **New**

 7. RP2040-based boards, such as **Nano RP2040 Connect**, or **RASPBERRY_PI_PICO_W with CYW43439 WiFi**, using [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed) or [**Earle Philhower's arduino-pico** core](https://github.com/earlephilhower/arduino-pico)
 
 8. RP2040-based boards, such as **RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040 and GENERIC_RP2040**, using [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed) or [**Earle Philhower's arduino-pico** core](https://github.com/earlephilhower/arduino-pico).

 9. **ESP32**
 
  - ESP32 boards, such as `ESP32_DEV`, etc.
  - ESP32S2-based boards, such as `ESP32S2_DEV`, `ESP32_S2 Saola`, etc.
  - ESP32C3-based boards, such as `ESP32C3_DEV`, etc.
  - ESP32_S3 (ESP32S3_DEV, ESP32_S3_BOX, UM TINYS3, UM PROS3, UM FEATHERS3, etc.)
  
10. **ESP8266**

11. **Portenta_H7**  **New**

12. **nRF52 boards**, such as **AdaFruit Feather nRF52832, nRF52840 Express, BlueFruit Sense, Itsy-Bitsy nRF52840 Express, Metro nRF52840 Express, NINA_B302_ublox, NINA_B112_ublox, etc.**

13. Arduino `megaAVR` boards using Arduino core

  - UNO WiFi Rev2   : `WiFi101`      **New**
  - Nano Every      : `WiFiEspAT`    **New**
  
14. CO2 Ampel

  - SAMD      : `WiFi101`      **New**

15. STM32 using `STM32duino Maple` core. **New**

16. Sparkfun SAMD

- SAMD21, SAMD51 : `WiFiEspAT`  **New**

17. Industruino SAMD

  - D21G : WiFiEspAT  **New**

18. Tlera Corp STM32WB boards

  - Firefly-WB55RG, Nucleo-WB55RG, etc. : `WiFiEspAT`  **New**

19. Maixduino boards

  - Sipeed Maixduino, etc. : `WiFiEspAT`  **New**

20. `RTL8720DN` using Realtek `AmebaD` core  **New**

21. Arduino, Sparkfun, Adafruit, etc. AVR boards (Mega, 32U4, etc.). To use patch for `ArduinoSTL` library **New**

  - Arduino Uno / Mega / Duemilanove / Diecimila / LilyPad / Mini / Fio / Nano, etc.
  - **Arduino ATMega 16U4, 32U4** such as AVR Leonardo, Leonardo ETH, YUN, Esplora, LILYPAD_USB, AVR_ROBOT_CONTROL, AVR_ROBOT_MOTOR, AVR_INDUSTRIAL101, etc.
  - **Adafruit ATMega 32U4** such as AVR_FLORA8, AVR_FEATHER32U4, AVR_CIRCUITPLAY, AVR_ITSYBITSY32U4_5V, AVR_ITSYBITSY32U4_3V, AVR_BLUEFRUITMICRO, AVR_ADAFRUIT32U4, etc.
  - **Adafruit ATMega 328(P)** such as AVR_METRO, AVR_FEATHER328P, AVR_PROTRINKET5, AVR_PROTRINKET3, AVR_PROTRINKET5FTDI, AVR_PROTRINKET3FTDI, etc.
  - **Generic or Sparkfun AVR ATmega_32U4** such as **AVR_MAKEYMAKEY, AVR_PROMICRO, etc.**
  - **Generic or Sparkfun AVR ATmega_328(P)** such as **ARDUINO_REDBOT, ARDUINO_AVR_DIGITAL_SANDBOX, etc.**
  - **Generic or Sparkfun AVR ATmega128RFA1** such as **ATMEGA128RFA1_DEV_BOARD, etc.**


---

#### Currently supported WiFi shields/modules

1. WiFiNINA using [`WiFiNINA_Generic library`](https://github.com/khoih-prog/WiFiNINA_Generic)
2. WiFi101 using [`WiFi101_Generic library`](https://github.com/khoih-prog/WiFi101_Generic)  **New**
3. u-blox W101, W102 using [`WiFiNINA_Generic library`](https://github.com/khoih-prog/WiFiNINA_Generic)
4. ESP8266-AT command using [`WiFiEspAT library`](https://github.com/jandrassy/WiFiEspAT)
5. ESP8266/ESP32-AT command using [`ESP_AT_Lib library`](https://github.com/khoih-prog/ESP_AT_Lib)
6. Built-in WiFi of ESP32, ESP8266
7. Built-in WiFi of Portenta_H7 **New**
8. Built-in CYW43439 WiFi of RASPBERRY_PI_PICO_W **New**

---
---

## Prerequisites

 1. [`Arduino IDE 1.8.19+` for Arduino](https://github.com/arduino/Arduino). [![GitHub release](https://img.shields.io/github/release/arduino/Arduino.svg)](https://github.com/arduino/Arduino/releases/latest)
 2. [`ESP32 Core 2.0.5+`](https://github.com/espressif/arduino-esp32) for ESP32-based boards. [![Latest release](https://img.shields.io/github/release/espressif/arduino-esp32.svg)](https://github.com/espressif/arduino-esp32/releases/latest/)
 3. [`ESP8266 Core 3.0.2+`](https://github.com/esp8266/Arduino) for ESP8266-based boards. [![Latest release](https://img.shields.io/github/release/esp8266/Arduino.svg)](https://github.com/esp8266/Arduino/releases/latest/).
 4. [`Arduino AVR core 1.8.6+`](https://github.com/arduino/ArduinoCore-avr) for Arduino (Use Arduino Board Manager) for AVR boards. [![GitHub release](https://img.shields.io/github/release/arduino/ArduinoCore-avr.svg)](https://github.com/arduino/ArduinoCore-avr/releases/latest) **New**
 5. [`Adafruit AVR core 1.4.15+`](https://github.com/adafruit/Adafruit_Arduino_Boards) for Adafruit AVR boards. Use Arduino Board Manager to install. [![GitHub release](https://img.shields.io/github/release/adafruit/Adafruit_Arduino_Boards.svg)](https://github.com/adafruit/Adafruit_Arduino_Boards/releases/latest) **New**
 6. [`Sparkfun AVR core 1.1.13+`](https://github.com/sparkfun/Arduino_Boards) for Sparkfun AVR boards. Use Arduino Board Manager to install. **New**
 7. [`Teensy core v1.57+`](https://www.pjrc.com/teensy/td_download.html) for Teensy (4.1, 4.0, 3.6, 3.5, 3,2, 3.1, 3.0) boards.
 8. [`Arduino SAM DUE core v1.6.12+`](https://github.com/arduino/ArduinoCore-sam) for SAM DUE ARM Cortex-M3 boards.
 9. [`Arduino SAMD core 1.8.13+`](https://github.com/arduino/ArduinoCore-samd) for SAMD ARM Cortex-M0+ boards. [![GitHub release](https://img.shields.io/github/release/arduino/ArduinoCore-samd.svg)](https://github.com/arduino/ArduinoCore-samd/releases/latest)
10. [`Adafruit SAMD core 1.7.11+`](https://github.com/adafruit/ArduinoCore-samd) for SAMD ARM Cortex-M0+ and M4 boards (Nano 33 IoT, etc.). [![GitHub release](https://img.shields.io/github/release/adafruit/ArduinoCore-samd.svg)](https://github.com/adafruit/ArduinoCore-samd/releases/latest)
11. [`Seeeduino SAMD core 1.8.3+`](https://github.com/Seeed-Studio/ArduinoCore-samd) for SAMD21/SAMD51 boards (XIAO M0, Wio Terminal, etc.). [![Latest release](https://img.shields.io/github/release/Seeed-Studio/ArduinoCore-samd.svg)](https://github.com/Seeed-Studio/ArduinoCore-samd/releases/latest/)
12. [`Adafruit nRF52 v1.3.0+`](https://github.com/adafruit/Adafruit_nRF52_Arduino) for nRF52 boards such as Adafruit NRF52840_FEATHER, NRF52832_FEATHER, NRF52840_FEATHER_SENSE, NRF52840_ITSYBITSY, NRF52840_CIRCUITPLAY, NRF52840_CLUE, NRF52840_METRO, NRF52840_PCA10056, PARTICLE_XENON, **NINA_B302_ublox**, etc. [![GitHub release](https://img.shields.io/github/release/adafruit/Adafruit_nRF52_Arduino.svg)](https://github.com/adafruit/Adafruit_nRF52_Arduino/releases/latest)
13. [`Arduino Core for STM32 v2.3.0+`](https://github.com/stm32duino/Arduino_Core_STM32) for STM32F/L/H/G/WB/MP1 boards. [![GitHub release](https://img.shields.io/github/release/stm32duino/Arduino_Core_STM32.svg)](https://github.com/stm32duino/Arduino_Core_STM32/releases/latest)
14. [`Earle Philhower's arduino-pico core v2.6.3+`](https://github.com/earlephilhower/arduino-pico) for RP2040-based boards such as **RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040 and GENERIC_RP2040**, etc. [![GitHub release](https://img.shields.io/github/release/earlephilhower/arduino-pico.svg)](https://github.com/earlephilhower/arduino-pico/releases/latest)
15. [`ArduinoCore-mbed mbed_rp2040, mbed_nano, mbed_portenta core 3.4.1+`](https://github.com/arduino/ArduinoCore-mbed) for Arduino (Use Arduino Board Manager) **Portenta_H7, RP2040-based boards, such as Nano_RP2040_Connect, RASPBERRY_PI_PICO**. [![GitHub release](https://img.shields.io/github/release/arduino/ArduinoCore-mbed.svg)](https://github.com/arduino/ArduinoCore-mbed/releases/latest)
16. [`Arduino megaAVR core 1.8.7+`](https://github.com/arduino/ArduinoCore-megaavr/releases) for Arduino megaAVR boards such as **Arduino UNO WiFi Rev2, AVR_NANO_EVERY, etc.**
17. [`STM32 Arduino_STM32 core 1.0.0+`](https://github.com/rogerclarkmelbourne/Arduino_STM32) for STM32F boards.  [![GitHub release](https://img.shields.io/github/release/rogerclarkmelbourne/Arduino_STM32.svg)](https://github.com/rogerclarkmelbourne/Arduino_STM32/releases/latest). **New**
18. [`MegaCoreX megaAVR core 1.1.0+`](https://github.com/MCUdude/MegaCoreX/releases) for Arduino megaAVR boards.  [![GitHub release](https://img.shields.io/github/release/MCUdude/MegaCoreX.svg)](https://github.com/MCUdude/MegaCoreX/releases/latest). Follow [**How to install**](https://github.com/MCUdude/MegaCoreX#how-to-install). **Not yet**
19. [`Functional-Vlpp library v1.0.2+`](https://github.com/khoih-prog/functional-vlpp) to use server's lambda function. To install. check [![arduino-library-badge](https://www.ardu-badge.com/badge/Functional-Vlpp.svg?)](https://www.ardu-badge.com/Functional-Vlpp)
20. [`WiFiNINA_Generic library v1.8.15-1+`](https://github.com/khoih-prog/WiFiNINA_Generic) if using WiFiNINA. To install. check [![arduino-library-badge](https://www.ardu-badge.com/badge/WiFiNINA_Generic.svg?)](https://www.ardu-badge.com/WiFiNINA_Generic).
21. [`WiFi101_Generic Library v1.0.0+`](https://github.com/khoih-prog/WiFi101_Generic) to use SAMD MKR1000, etc. boards with WiFi101 to send large data. To install. check [![arduino-library-badge](https://www.ardu-badge.com/badge/WiFi101_Generic.svg?)](https://www.ardu-badge.com/WiFi101_Generic).
22. [`WiFiEspAT library v1.4.1+`](https://github.com/jandrassy/WiFiEspAT) if using ESP8288/ESP32-AT shields. [![GitHub release](https://img.shields.io/github/release/jandrassy/WiFiEspAT.svg)](https://github.com/jandrassy/WiFiEspAT/releases/latest)
23. [`ArduinoSTL library v1.3.3+`](https://github.com/mike-matera/ArduinoSTL) if using AVR boards. [![GitHub release](https://img.shields.io/github/release/mike-matera/ArduinoSTL.svg)](https://github.com/mike-matera/ArduinoSTL/releases/latest)
24. [`ESP_AT_Lib library v1.4.1+`](https://github.com/khoih-prog/ESP_AT_Lib) if using ESP8288/ESP32-AT shields. To install, check [![arduino-library-badge](https://www.ardu-badge.com/badge/ESP_AT_Lib.svg?)](https://www.ardu-badge.com/ESP_AT_Lib).

---
---

## Installation

### Use Arduino Library Manager

The best and easiest way is to use `Arduino Library Manager`. Search for `WiFiMulti_Generic`, then select / install the latest version.
You can also use this link [![arduino-library-badge](https://www.ardu-badge.com/badge/WiFiMulti_Generic.svg?)](https://www.ardu-badge.com/WiFiMulti_Generic) for more detailed instructions.

### Manual Install

1. Navigate to [WiFiMulti_Generic](https://github.com/khoih-prog/WiFiMulti_Generic) page.
2. Download the latest release `WiFiMulti_Generic-main.zip`.
3. Extract the zip file to `WiFiMulti_Generic-main` directory 

changelog.md
## WiFiMulti_Generic Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/WiFiMulti_Generic.svg?)](https://www.ardu-badge.com/WiFiMulti_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/WiFiMulti_Generic.svg)](https://github.com/khoih-prog/WiFiMulti_Generic/releases)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/WiFiMulti_Generic.svg)](http://github.com/khoih-prog/WiFiMulti_Generic/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of Contents


* [Changelog](#changelog)
  * [Releases v1.2.2](#releases-v122)
  * [Releases v1.2.1](#releases-v121)
  * [Releases v1.2.0](#releases-v120)
  * [Releases v1.1.1](#releases-v111)
  * [Releases v1.1.0](#releases-v110)
  * [Releases v1.0.0](#releases-v100)

---
---

## Changelog

#### Releases v1.2.2

1.  Better workaround for RP2040W WiFi.status() bug using `ping()` to local gateway. Check [WiFi.status() wrongly reports WL_CONNECTED even when WiFi is lost and RSSI is always 0 dBm #762](https://github.com/earlephilhower/arduino-pico/issues/762)

#### Releases v1.2.1

1. Workaround for RP2040W WiFi.status() bug. Check [WiFi.status() wrongly reports WL_CONNECTED even when WiFi is lost and RSSI is always 0 dBm #762](https://github.com/earlephilhower/arduino-pico/issues/762)
2. Add example [WiFiMulti_RP2040W](examples/WiFiMulti_RP2040W)

#### Releases v1.2.0

1. Add support to RASPBERRY_PI_PICO_W using CYW43439 WiFi
2. Update `Packages_Patches`

#### Releases v1.1.1

1. Fix bug
2. Update `Packages_Patches`

#### Releases v1.1.0

1. Add support to 

  - **Portenta_H7 using Murata WiFi**
  - **RTL8720DN using built-in 2.4GHz or 5GHz WiFi** with Realtek `AmebaD` core
  - Arduino SAMD using `WiFi101` (MKR1000, MKR1010, etc.)
  - Seeed XIAO RP2020, SAMD using `WiFi101`, `WiFiNINA_Generic` or `WiFiEspAT`
  - Arduino megaAVR boards with Arduino core (UNO WiFi Rev2 using `WiFi101`, Nano Every using `WiFi101`, `WiFiNINA_Generic` or `WiFiEspAT`)
  - CO2 Ampel SAMD board using `WiFi101`, `WiFiNINA_Generic` or `WiFiEspAT`
  - STM32 boards with `STM32duino Maple` core using `WiFi101`, `WiFiNINA_Generic` or `WiFiEspAT`
  - Sparkfun SAMD board using `WiFi101`, `WiFiNINA_Generic` or `WiFiEspAT`
  - Industruino SAMD using `WiFi101`, `WiFiNINA_Generic` or `WiFiEspAT`
  - Tlera Corp STM32WB boards (Firefly-WB55RG, Nucleo-WB55RG, etc.) using `WiFi101`, `WiFiNINA_Generic` or `WiFiEspAT`
  - Maixduino boards using `WiFi101`, `WiFiNINA_Generic` or `WiFiEspAT`
  - Arduino, Sparkfun, Adafruit, etc. AVR boards (Mega, 32U4, etc.). To use `LibraryPatches` for `ArduinoSTL` library

2. Update example [WiFiMulti](examples/WiFiMulti) to handle more complex cases.


#### Releases v1.0.0

1. Initial coding to support **ESP32, ESP8266, WiFiNINA and ESP8266_AT and ESP32_AT WiFi** for many boards (nRF52, SAMD, Teensy, RP2040, SAM-DUE, ESP32, ESP8266, etc.)





src/ESP32/WiFiMulti_Generic_Impl.h
/**************************************************************************************************************************************
  WiFiMulti_Generic_Impl.h
  For any WiFi shields, such as ESP32, ESP8266, Portenta_H7, WiFiNINA W101, W102, W13x, or custom, such as ESP8266/ESP32-AT, etc

  WiFiMulti_Generic is a library to adapt the  ESP32/ESP8266 WiFiMulti feature to other WiFi modules

  Based on and modified from WiFiMulti of ESP32 core: https://github.com/espressif/arduino-esp32
  Based on and modified from WiFiMulti of ESP8266 core: https://github.com/esp8266/Arduino

  Built by Khoi Hoang https://github.com/khoih-prog/WiFiMulti_Generic

  License under GPL-3.0

  This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License
  as published bythe Free Software Foundation, either version 3 of the License, or (at your option) any later version.
  This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
  You should have received a copy of the GNU General Public License along with this program.  If not, see <https://www.gnu.org/licenses/>

  Original author:
  @file WiFiMulti.cpp
  @date 16.05.2015
  @author Markus Sattler
  Copyright (c) 2015 Markus Sattler. All rights reserved.
  This file is part of the esp32 core for Arduino environment.

  Version: 1.2.2

  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  1.0.0   K Hoang      15/02/2020 Initial coding for ESP32, ESP8266, WiFiNINA and ESP_AT modules
  1.1.0   K Hoang      16/02/2020 Add support to WiFi101 and many more boards (PortentaH7, megaAVR, Sparkfun SAMD, etc.)
  1.1.1   K Hoang      26/04/2020 Fix bug
  1.2.0   K Hoang      12/08/2022 Add support to RASPBERRY_PI_PICO_W using CYW4343 WiFi
  1.2.1   K Hoang      16/08/2022 Workaround for RP2040W WiFi.status() bug
  1.2.2   K Hoang      16/08/2022 Better workaround for RP2040W WiFi.status() bug using ping() to local gateway
 ***************************************************************************************************************************************/

#pragma once

#ifndef _WIFIMULTI_GENERIC_IMPL_H_
#define _WIFIMULTI_GENERIC_IMPL_H_

#include <limits.h>
#include <string.h>

WiFiMulti_Generic::WiFiMulti_Generic()
{
}

WiFiMulti_Generic::~WiFiMulti_Generic()
{
  for (uint32_t i = 0; i < APlist.size(); i++)
  {
    WifiAPlist_t entry = APlist[i];

    if (entry.ssid)
    {
      free(entry.ssid);
    }

    if (entry.passphrase)
    {
      free(entry.passphrase);
    }
  }

  APlist.clear();
}

bool WiFiMulti_Generic::addAP(const char* ssid, const char *passphrase)
{
  WifiAPlist_t newAP;

  if (!ssid || *ssid == 0x00 || strlen(ssid) > 31)
  {
    // fail SSID too long or missing!
    WFM_LOGERROR("[addAP] No ssid or ssid too long");
    return false;
  }

  if (passphrase && strlen(passphrase) > 64)
  {
    // fail passphrase too long!
    WFM_LOGERROR("[addAP] Passphrase too long");
    return false;
  }

  newAP.ssid = strdup(ssid);

  if (!newAP.ssid)
  {
    WFM_LOGERROR("[addAP] Fail : newAP.ssid == 0");
    return false;
  }

  if (passphrase && *passphrase != 0x00)
  {
    newAP.passphrase = strdup(passphrase);

    if (!newAP.passphrase)
    {
      WFM_LOGERROR("[addAP] Fail : newAP.passphrase == 0");
      free(newAP.ssid);
      return false;
    }
  }
  else
  {
    newAP.passphrase = NULL;
  }

  APlist.push_back(newAP);
  WFM_LOGINFO1("[addAP] add SSID:", newAP.ssid);

  return true;
}

uint8_t WiFiMulti_Generic::run(const uint32_t& connectTimeout)
{
  int8_t scanResult;
  uint8_t status = WiFi.status();

  if (status == WL_CONNECTED)
  {
    for (uint32_t x = 0; x < APlist.size(); x++)
    {
      if (WiFi.SSID() == APlist[x].ssid)
      {
        return status;
      }
    }

    WiFi.disconnect(false, false);

    delay(10);
    status = WiFi.status();
  }

  scanResult = WiFi.scanNetworks();

  if (scanResult == WIFI_SCAN_RUNNING)
  {
    // scan is running
    return WL_NO_SSID_AVAIL;
  }
  else if (scanResult >= 0)
  {
    // scan done analyze
    WifiAPlist_t bestNetwork { NULL, NULL };

    int bestNetworkDb = INT_MIN;
    uint8_t bestBSSID[6];
    int32_t bestChannel = 0;

    WFM_LOGINFO("[run] Scan done");

    if (scanResult == 0)
    {
      WFM_LOGERROR("[run] No networks found");
    }
    else
    {
      WFM_LOGINFO1("[run] Number of Networks found:", scanResult);

      for (int8_t i = 0; i < scanResult; ++i)
      {
        String ssid_scan;
        int32_t rssi_scan;
        uint8_t sec_scan;
        uint8_t* BSSID_scan;
        int32_t chan_scan;

        WiFi.getNetworkInfo(i, ssid_scan, sec_scan, rssi_scan, BSSID_scan, chan_scan);

        bool known = false;

        for (uint32_t x = APlist.size() ; x > 0; x--)
        {
          WifiAPlist_t entry = APlist[x - 1];

          if (ssid_scan == entry.ssid)
          {
            // SSID match
            known = true;

            if (rssi_scan > bestNetworkDb)
            {
              // best network
              if (sec_scan == WIFI_AUTH_OPEN || entry.passphrase)
              {
                // check for passphrase if not open wlan
                bestNetworkDb = rssi_scan;
                bestChannel = chan_scan;
                memcpy((void*) &bestNetwork, (void*) &entry, sizeof(bestNetwork));
                memcpy((void*) &bestBSSID, (void*) BSSID_scan, sizeof(bestBSSID));
              }
            }

            break;
          }
        }

        WFM_PRINT_LINE;

        if (known)
        {
          WFM_LOGDEBUG3("  Known => #", i, ", Channel:", chan_scan);
          WFM_HEXLOGDEBUG5_SEPARATOR(":", BSSID_scan[0], BSSID_scan[1], BSSID_scan[2], BSSID_scan[3], BSSID_scan[4],
                                     BSSID_scan[5]);
          WFM_LOGDEBUG5("SSID:", ssid_scan.c_str(), ", RSSI:", rssi_scan, ", Secured:", (sec_scan == WIFI_AUTH_OPEN) ? 'n' : 'y');
        }
        else
        {
          WFM_LOGDEBUG3("Unknown => #", i, ", Channel:", chan_scan);
          WFM_HEXLOGDEBUG5_SEPARATOR(":", BSSID_scan[0], BSSID_scan[1], BSSID_scan[2], BSSID_scan[3], BSSID_scan[4],
                                     BSSID_scan[5]);
          WFM_LOGDEBUG5("SSID:", ssid_scan.c_str(), ", RSSI:", rssi_scan, ", Secured:", (sec_scan == WIFI_AUTH_OPEN) ? 'n' : 'y');
        }
      }
    }

    WFM_PRINT_LINE;

    // clean up ram
    WiFi.scanDelete();

    if (bestNetwork.ssid)
    {
      WFM_LOGINFO0("[run] Connecting BSSID: ");
      WFM_HEXLOGINFO5_SEPARATOR(":", bestBSSID[0], bestBSSID[1], bestBSSID[2], bestBSSID[3], bestBSSID[4], bestBSSID[5]);
      WFM_LOGINFO5("SSID: ", bestNetwork.ssid, ", Channel: ", bestChannel, ", Best dB: ", bestNetworkDb);

      WiFi.begin(bestNetwork.ssid, bestNetwork.passphrase, bestChannel, bestBSSID);
      status = WiFi.status();

      auto startTime = millis();

      // wait for connection, fail, or timeout
      while (status != WL_CONNECTED && status != WL_NO_SSID_AVAIL && status != WL_CONNECT_FAILED
             && (millis() - startTime) <= connectTimeout)
      {
        delay(10);
        status = WiFi.status();
      }

      switch (status)
      {
        case WL_CONNECTED:
          WFM_LOGINFO("[run] Connecting done.");
          WFM_LOGDEBUG3("[run] SSID: ", WiFi.SSID().c_str(), ", IP: ", WiFi.localIP().toString().c_str());
          WFM_LOGDEBUG3("[run] MAC: ",  WiFi.BSSIDstr().c_str(), ", Channel: ", WiFi.channel());

          break;

        case WL_NO_SSID_AVAIL:
          WFM_LOGERROR("[run] Connecting Failed, AP not found.");
          break;

        case WL_CONNECT_FAILED:
          WFM_LOGERROR("[run] Connecting Failed.");
          break;

        default:
          WFM_LOGERROR1("[run] Connecting Failed, status =", status);
          break;
      }
    }
    else
    {
      WFM_LOGERROR("[run] no matching wifi found!");
    }
  }
  else
  {
    // start scan
    WFM_LOGDEBUG("[run] delete old wifi config...");
    WiFi.disconnect();

    WFM_LOGDEBUG("[run] start scan");
    // scan wifi async mode
    WiFi.scanNetworks(true);
  }

  return status;
}

#endif    // _WIFIMULTI_GENERIC_IMPL_H_
